pybaseball/split_stats.py

```diff
@@ -0,0 +1,116 @@
+import requests
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+import bs4 as bs
+import pandas as pd
+import re
+
+def download_url(url):
+    """
+    Gets the content from the url specified
+    """
+    session = requests.Session()
+    retry = Retry(connect=3, backoff_factor=0.5)
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount('http://', adapter)
+    session.mount('https://', adapter)
+    resp=session.get(url)
+    return resp.content
+
+def get_split_soup(playerid, year=None, pitching_splits=False):
+    """
+    gets soup for the player splits. 
+    """    
+    if pitching_splits is True:
+        pitch_or_bat='p'
+    else:
+        pitch_or_bat='b'
+    if year is None: #provides scores from yesterday if date is not provided
+        url=f"https://www.baseball-reference.com/players/split.fcgi?id={playerid}&year=Career&t={pitch_or_bat}"
+    else:
+        year=str(year)
+        url=f"https://www.baseball-reference.com/players/split.fcgi?id={playerid}&year={year}&t={pitch_or_bat}"
+    html=download_url(url)
+    soup=bs.BeautifulSoup(html, 'lxml')
+    return soup
+
+def get_player_info(playerid=None, soup=None):
```

Comment:
I think this would be a lot simpler if you used the `HTMLTableProcessor` class in `datasources/html_table_processor.py` along with xpath. However I also understand all the other BRef calls work this way right now.

Reply:
I haven't tried using the HTMLTableProcessor class, though it looks pretty helpful for this sort of thing. The tricky part about this bbref page is that the table for the splits is wrapped in comments and the player info was all within <p> tags. I had to do an extra layer of identifying the comments and then parsing for the table to get to the result. Again, haven't tried using this class, but it seems unlikely to work because of that formatting.

Reply:
Xpath is pretty powerful, and supports comments, so it should work. I'll play around with it and see.

Reply:
It is admittedly a LITTLE more complicated than I'd like, but this would replace most of `split_stats` and could be adapted for `get_player_info`.

I will preface this by saying, don't add this if you don't feel comfortable. I plan on refactoring a lot of the BRef code at some point, so this would just be another one on the list. 😆 

As well if you DO choose to use this, feel free to move the xpath addon functions to the `html_table_processor.py`

```python
import re
from typing import Any, List, TypeVar, Union

import lxml.etree
import pandas as pd
import requests

from pybaseball.datasources.html_table_processor import HTMLTableProcessor

_T = TypeVar('_T')
ListOrSelf = Union[_T, List[_T]]

@lxml.etree.FunctionNamespace(None)
def parse(context: Any, html_str: str) -> lxml.etree._Element:
    return lxml.etree.HTML(html_str).xpath('/html/body/*')[0]
    # return lxml.etree.fromstring(html_str)

@lxml.etree.FunctionNamespace(None)
def parse_comment(context: Any, html_comment: ListOrSelf[lxml.etree._Comment]) -> ListOrSelf[lxml.etree._Element]:
    if isinstance(html_comment, list):
        return [parse_comment(context, x) for x in html_comment]
    elif isinstance(html_comment, lxml.etree._Comment):
        html_str = lxml.etree.tostring(html_comment).decode('utf-8')
        html_str = html_str.replace("<!--", "").replace("-->", "")
        return parse(context, html_str)
    else:
        raise ValueError(f"Expected a Comment or List[Comment], got {type(html_comment)}")

@lxml.etree.FunctionNamespace(None)
def safe_text(context: Any, element: ListOrSelf[lxml.etree._Element], default: str = '') -> ListOrSelf[str]:
    if isinstance(element, list):
        return [safe_text(context, x, default) for x in element]
    elif isinstance(element, lxml.etree._Element):
        return element.text or default
    else:
        raise ValueError(f"Expected an Element or List[Element], got {type(element)}")

playerid = 'troutmi01'
pitch_or_bat = 'b'
year = 2019

bref_html = requests.get(f"https://www.baseball-reference.com/players/split.fcgi?id={playerid}&year={year}&t={pitch_or_bat}").content.decode('utf-8')

bref_comment_tables = [
    # Why do you hurt me so lxml?
    lxml.etree.fromstring(lxml.etree.tostring(x)) for x in
    lxml.etree.HTML(bref_html).xpath('parse_comment((//*[contains(comment(), "overthrow table_container")])//comment())')
]

root_url = "https://www.baseball-reference.com"
headings_xpath = "({TABLE_XPATH}/thead//th)/descendant-or-self::*/text()"
data_rows_xpath = "({TABLE_XPATH}/tbody//tr)"
data_cell_xpath = "safe_text((td|th)/descendant-or-self::*)"

processor = HTMLTableProcessor(root_url, headings_xpath, data_rows_xpath, data_cell_xpath, None)

split_tables = [processor.get_tabular_data_from_element(x) for x in bref_comment_tables]
for index, table in enumerate(split_tables):
    table['Split Type'] = bref_comment_tables[index].xpath('//table/caption/text()')[0]

data = pd.concat(split_tables).reset_index(drop=True)
data['1B'] = data['H'] - data['2B'] - data['3B'] - data['HR']
data['Player ID'] = playerid
data = data.set_index(['Player ID', 'Split Type', 'Split'])
data = data.drop(index=['Split'], level=2)
data = data.dropna(axis=1, how='all')

print(data)
```

Reply:
I would prefer to keep the original implementation on this. I like the consistency of what I have with the rest of the bbref parsing functions and honestly, I don't follow everything you have here.

Reply:
Completely understand.